### PR TITLE
Executor: Configurable checks on pods in state Pending 

### PR DESCRIPTION
**Pull request by robertdavidsmith**
When a pod is stuck in pending, we currently have hard-coded rules to determine pod whether to wait, hard-fail the job, or retry. 

This PR makes two changes

1) Replace the hard-coded checks with configurable checks.
2) Some changes to the checks (described below).

Old behavior:
If container state is InvalidImageName Fail after 90s
If event contains "desc = failed to pull and unpack image" or "code = Unknown desc = Error response from daemon" Fail after 90s
If there is a warning event other than "FailedScheduling" FAIL after stuck pod expiry (3m)
If container state is ImagePullBackOff or ErrImagePull FAIL after stuck pod expiry (3m)
Else RETRY after stuck pod expiry (3m)

New behavior:
If event contains "desc = failed to pull and unpack image" or "code = Unknown desc = Error response from daemon" FAIL after 90s
If there is a warning event other than "FailedScheduling" RETRY after 5m
If container state is InvalidImageName FAIL immediately
If container state is ImagePullBackOff or ErrImagePull RETRY after 15m
If any other container state RETRY if pending for 5m

Behavior change from before isn't that significant, there will almost certainly be more changes in later PRs.

### CODE REVIEW

File: internal/executor/podchecks/container_state_checks_test.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+	pod.Status.ContainerStatuses[0].State.Waiting = &v1.ContainerStateWaiting{Reason: "ImagePullBackOff", Message: "Backing off"}
+	pod.Status.ContainerStatuses[1].State.Waiting = &v1.ContainerStateWaiting{Reason: "InvalidImageName", Message: "Image name is wrong"}
+
+	action, message := csc.getAction(pod, time.Minute*2)
+	assert.Equal(t, ActionFail, action)
+	assert.NotEmpty(t, message)
+}
+
+func Test_getAction_WhenOneFailedContainer_FirstMatchingCheckIsNotTimedOut_ButLaterMatchingCheckIsTimedOut_Waits(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (robertdavidsmith) [reply internal/executor/podchecks/container_state_checks_test.go:91]: Yes, this is a deliberate choice. First rule to match identifies the state, then we apply the timeout for that state.

File: internal/executor/podchecks/poc_checks_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
+package podchecks
```
[the remark this answers is not in the document]
Comment (robertdavidsmith) [reply internal/executor/podchecks/poc_checks_test.go:1]: Will fix

File: internal/executor/podchecks/container_state_checks_test.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+	csc, err := newContainerStateChecks([]config.ContainerStatusCheck{})
+	assert.Nil(t, err)
+
+	action, message := csc.getAction(&v1.Pod{}, time.Second)
+	assert.Equal(t, ActionWait, action)
+	assert.Empty(t, message)
+}
+
+func Test_getAction_WhenOneFailedContainer_AndOneMatchingFailCheck_ReturnsFail(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (robertdavidsmith) [reply internal/executor/podchecks/container_state_checks_test.go:25]: Fair point will change

File: internal/executor/job/job_context.go [hunk 4]
```diff
@@ -212,30 +211,36 @@ func (c *ClusterJobContext) detectStuckPods(runningJob *RunningJob) {
...
+				message := createStuckPodMessage(retryable, podCheckMessage)
+
+				log.Warnf("Found issue with pod %s in namespace %s: %s", pod.Name, pod.Namespace, message)
+
+				c.registerIssue(runningJob, &PodIssue{
+					OriginatingPod: pod.DeepCopy(),
+					Pods:           runningJob.ActivePods,
+					Message:        message,
+					Retryable:      retryable,
+					Type:           UnableToSchedule,
```
[the remark this answers is not in the document]
Comment (robertdavidsmith) [reply internal/executor/job/job_context.go:240]: Reasonable, this PR is too big already, will do as a separate one

File: config/executor/config.yaml [hunk 5]
```diff
@@ -19,7 +19,37 @@ kubernetes:
...
+      action: Retry
+    containerStatuses:
+    - state: Waiting
+      reasonRegexp: "InvalidImageName"
+      timeout: 1s
+      action: Fail
+    - state: Waiting
+      reasonRegexp: "ImagePullBackOff"
+      timeout: 30m
+      action: Retry
```
[the remark this answers is not in the document]
Comment (robertdavidsmith) [reply config/executor/config.yaml:49]: I've made this 15m

File: internal/executor/podchecks/poc_checks_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (robertdavidsmith) [reply internal/executor/podchecks/poc_checks_test.go]: Fixed

File: internal/executor/podchecks/container_state_checks_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertdavidsmith) [reply internal/executor/podchecks/container_state_checks_test.go]: Fixed

File: config/executor/config.yaml [hunk 6]
```diff
@@ -19,7 +19,37 @@ kubernetes:
...
       hostnameSuffix: "svc"
+  pendingPodChecks:
+    events:
+    - regexp: "Failed to pull image.*desc = failed to pull and unpack image"
+      type: Warning
+      action: Fail
+    - regexp: "Failed to pull image.*code = Unknown desc = Error response from daemon"
+      type: Warning
+      action: Fail
+    - regexp: "nodes are available"  # match all except the FailedScheduling message "0/3 nodes are available:"
```
[the remark this answers is not in the document]
Comment (robertdavidsmith) [reply config/executor/config.yaml]: Setting grace period per event now, so timeout will be 5m

**Review by JamesMurkin — APPROVED** [review]

**Pull request merged**